# Projeto 1 - Ciência dos Dados

Nome: Ananda Júlia Galvão Campelo

Nome: Carol Magario de Mattos Barretto

Atenção: Serão permitidos grupos de três pessoas, mas com uma rubrica mais exigente. Grupos deste tamanho precisarão fazer um questionário de avaliação de trabalho em equipe

___
Carregando algumas bibliotecas:

In [49]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

In [50]:
print('Esperamos trabalhar no diretório')
print(os.getcwd())

Esperamos trabalhar no diretório
C:\Users\Ananda\CDADOS 22.2\22-2a-cd-p1-grupo_anandajgc


Carregando a base de dados com as notícias classificadas manualmente:

In [51]:
filename = 'data/dados.xlsx'

In [52]:
train = pd.read_excel(filename, sheet_name = 'Treinamento')
train.head(5)

Categoria                                             Titulo  \
0              Mercados   Ibovespa: ALPA4 fecha mais um pregão em alta e...   
1              Empresas   IRB (IRBR3) é, de longe, a ação mais ‘shortead...   
2  Carteira Recomendada   5 ações que pagam bons dividendos para investi...   
3  Carteira Recomendada   Empiricus Investimentos escala 6 ações com div...   
4              Mercados   Ibovespa futuro opera em queda com aversão ao ...   

                                       PrimeiroParag                Data  \
0  O Ibovespa (IBOV) fechou em queda de 1,09% nes...  26/08/2022 - 17:37   
1  O IRB (IRBR3) é a ação com mais investidores s...  31/08/2022 - 17:00   
2  Para setembro, a Órama Investimentos optou por...  05/09/2022 - 10:27   
3  A Empiricus Investimentos, ex-Vitreo, incluiu ...  05/09/2022 - 13:21   
4  O dólar abria o pregão desta segunda-feira (29...   29/08/2022 - 9:08   

   Pagina  Target  
0      43       2  
1      28       1  
2      14       1  
3      12       1  
4      40       0

In [53]:
test = pd.read_excel(filename, sheet_name = 'Teste')
test.head(5)

Categoria                                             Titulo  \
0              Mercados   Ibovespa (IBOV) vira para queda, com VALE3 ent...   
1              Empresas   Ação da Braskem (BRKM3) avança com notícia sob...   
2              Mercados   Compre commodities, se preocupe com recessão d...   
3  Carteira Recomendada   Veja as 5 novas ações escaladas pela XP para e...   
4            Mais Lidas   Respostas sobre Magazine Luiza (MGLU3) e mais ...   

                                       PrimeiroParag                Data  \
0  O Ibovespa (IBOV) operava sem forças nesta qua...  24/08/2022 - 13:58   
1  As ações da Braskem (BRKM3) avançavam ao redor...  02/09/2022 - 11:30   
2  O Goldman Sachs recomenda que os investidores ...   29/08/2022 - 9:49   
3  A XP Investimentos repaginou sua carteira reco...   20/08/2022 - 9:31   
4  Investidores não tiraram Magazine Luiza (MGLU3...  28/08/2022 - 10:30   

   Pagina  Target  
0    53.0       2  
1    20.0       2  
2    40.0       1  
3    66.0       1  
4    42.0       2

In [54]:
test.drop(['Categoria','Data','Pagina'], axis = 1, inplace = True)
test

Titulo  \
0    Ibovespa (IBOV) vira para queda, com VALE3 ent...   
1    Ação da Braskem (BRKM3) avança com notícia sob...   
2    Compre commodities, se preocupe com recessão d...   
3    Veja as 5 novas ações escaladas pela XP para e...   
4    Respostas sobre Magazine Luiza (MGLU3) e mais ...   
..                                                 ...   
295  Ibovespa (IBOV) fecha em alta e recupera os 11...   
296  Ambev (ABEV3) está com desconto de 30%, calcul...   
297  brMalls (BRML3), Gerdau (GGBR4) e mais: Veja a...   
298  TSE homologa candidatura de Bolsonaro; Tribuna...   
299  Ibovespa (IBOV) reduz ganhos com Wall Street, ...   

                                         PrimeiroParag  Target  
0    O Ibovespa (IBOV) operava sem forças nesta qua...       2  
1    As ações da Braskem (BRKM3) avançavam ao redor...       2  
2    O Goldman Sachs recomenda que os investidores ...       1  
3    A XP Investimentos repaginou sua carteira reco...       1  
4    Investidores não tiraram Magazine Luiza (MGLU3...       2  
..                                                 ...     ...  
295  O Ibovespa (IBOV) fechou em alta nesta quinta-...       1  
296  A Ambev (ABEV3) está com desconto de 30% sobre...       2  
297  Em atualização de carteira semanal, a XP Inves...       1  
298  Na sessão de julgamentos desta terça (6), os m...       0  
299  O Ibovespa (IBOV) avançou nesta sexta-feira (2...       1  

[300 rows x 3 columns]

In [55]:
train.drop(['Categoria','Data','Pagina'], axis = 1, inplace = True)
train

Titulo  \
0    Ibovespa: ALPA4 fecha mais um pregão em alta e...   
1    IRB (IRBR3) é, de longe, a ação mais ‘shortead...   
2    5 ações que pagam bons dividendos para investi...   
3    Empiricus Investimentos escala 6 ações com div...   
4    Ibovespa futuro opera em queda com aversão ao ...   
..                                                 ...   
445  “Vamos lançar um NFT nos próximos meses que va...   
446     Digitalks terá presença de DAO 100% brasileira   
447  Vale (VALE3), Santander (SANB11) e mais 3 açõe...   
448  Credores da chinesa Evergrande (EGRNF) montam ...   
449  Ibovespa: PCAR3 dispara mais de 8% e lidera as...   

                                         PrimeiroParag  Target  
0    O Ibovespa (IBOV) fechou em queda de 1,09% nes...       2  
1    O IRB (IRBR3) é a ação com mais investidores s...       1  
2    Para setembro, a Órama Investimentos optou por...       1  
3    A Empiricus Investimentos, ex-Vitreo, incluiu ...       1  
4    O dólar abria o pregão desta segunda-feira (29...       0  
..                                                 ...     ...  
445  Neste sábado (3) aconteceu o ViralCure Festiva...       0  
446  Cintia Ferreira, Fundadora da EVE e co-fundado...       0  
447  A Ativa Investimentos retirou as ações de Bras...       2  
448  Fundos globais que investiram em títulos da ch...       1  
449  Confira as maiores altas e baixas do Ibovespa ...       2  

[450 rows x 3 columns]

In [56]:
# Nova variável
test['Texto Final']  = test['Titulo'] + ' ' + test['PrimeiroParag']
test

Titulo  \
0    Ibovespa (IBOV) vira para queda, com VALE3 ent...   
1    Ação da Braskem (BRKM3) avança com notícia sob...   
2    Compre commodities, se preocupe com recessão d...   
3    Veja as 5 novas ações escaladas pela XP para e...   
4    Respostas sobre Magazine Luiza (MGLU3) e mais ...   
..                                                 ...   
295  Ibovespa (IBOV) fecha em alta e recupera os 11...   
296  Ambev (ABEV3) está com desconto de 30%, calcul...   
297  brMalls (BRML3), Gerdau (GGBR4) e mais: Veja a...   
298  TSE homologa candidatura de Bolsonaro; Tribuna...   
299  Ibovespa (IBOV) reduz ganhos com Wall Street, ...   

                                         PrimeiroParag  Target  \
0    O Ibovespa (IBOV) operava sem forças nesta qua...       2   
1    As ações da Braskem (BRKM3) avançavam ao redor...       2   
2    O Goldman Sachs recomenda que os investidores ...       1   
3    A XP Investimentos repaginou sua carteira reco...       1   
4    Investidores não tiraram Magazine Luiza (MGLU3...       2   
..                                                 ...     ...   
295  O Ibovespa (IBOV) fechou em alta nesta quinta-...       1   
296  A Ambev (ABEV3) está com desconto de 30% sobre...       2   
297  Em atualização de carteira semanal, a XP Inves...       1   
298  Na sessão de julgamentos desta terça (6), os m...       0   
299  O Ibovespa (IBOV) avançou nesta sexta-feira (2...       1   

                                           Texto Final  
0    Ibovespa (IBOV) vira para queda, com VALE3 ent...  
1    Ação da Braskem (BRKM3) avança com notícia sob...  
2    Compre commodities, se preocupe com recessão d...  
3    Veja as 5 novas ações escaladas pela XP para e...  
4    Respostas sobre Magazine Luiza (MGLU3) e mais ...  
..                                                 ...  
295  Ibovespa (IBOV) fecha em alta e recupera os 11...  
296  Ambev (ABEV3) está com desconto de 30%, calcul...  
297  brMalls (BRML3), Gerdau (GGBR4) e mais: Veja a...  
298  TSE homologa candidatura de Bolsonaro; Tribuna...  
299  Ibovespa (IBOV) reduz ganhos com Wall Street, ...  

[300 rows x 4 columns]

In [57]:
# Nova variável
train['Texto Final']  = train['Titulo'] + ' ' + train['PrimeiroParag']
train

Titulo  \
0    Ibovespa: ALPA4 fecha mais um pregão em alta e...   
1    IRB (IRBR3) é, de longe, a ação mais ‘shortead...   
2    5 ações que pagam bons dividendos para investi...   
3    Empiricus Investimentos escala 6 ações com div...   
4    Ibovespa futuro opera em queda com aversão ao ...   
..                                                 ...   
445  “Vamos lançar um NFT nos próximos meses que va...   
446     Digitalks terá presença de DAO 100% brasileira   
447  Vale (VALE3), Santander (SANB11) e mais 3 açõe...   
448  Credores da chinesa Evergrande (EGRNF) montam ...   
449  Ibovespa: PCAR3 dispara mais de 8% e lidera as...   

                                         PrimeiroParag  Target  \
0    O Ibovespa (IBOV) fechou em queda de 1,09% nes...       2   
1    O IRB (IRBR3) é a ação com mais investidores s...       1   
2    Para setembro, a Órama Investimentos optou por...       1   
3    A Empiricus Investimentos, ex-Vitreo, incluiu ...       1   
4    O dólar abria o pregão desta segunda-feira (29...       0   
..                                                 ...     ...   
445  Neste sábado (3) aconteceu o ViralCure Festiva...       0   
446  Cintia Ferreira, Fundadora da EVE e co-fundado...       0   
447  A Ativa Investimentos retirou as ações de Bras...       2   
448  Fundos globais que investiram em títulos da ch...       1   
449  Confira as maiores altas e baixas do Ibovespa ...       2   

                                           Texto Final  
0    Ibovespa: ALPA4 fecha mais um pregão em alta e...  
1    IRB (IRBR3) é, de longe, a ação mais ‘shortead...  
2    5 ações que pagam bons dividendos para investi...  
3    Empiricus Investimentos escala 6 ações com div...  
4    Ibovespa futuro opera em queda com aversão ao ...  
..                                                 ...  
445  “Vamos lançar um NFT nos próximos meses que va...  
446  Digitalks terá presença de DAO 100% brasileira...  
447  Vale (VALE3), Santander (SANB11) e mais 3 açõe...  
448  Credores da chinesa Evergrande (EGRNF) montam ...  
449  Ibovespa: PCAR3 dispara mais de 8% e lidera as...  

[450 rows x 4 columns]

In [58]:
target = list(train['Target'])

In [59]:
train0 = train.loc[train['Target'] == 0,:]
train0

Titulo  \
4    Ibovespa futuro opera em queda com aversão ao ...   
8    Prisma Capital se torna titular de 56% da Domm...   
12   Day Trade: Banco do Brasil (BBAS3), Marfrig (M...   
14   Diretores da Aneel aprovam novas tarifas de en...   
15   Bolsonaro no JN: centrão, pandemia, economia; ...   
..                                                 ...   
439  Magalu (MGLU3) entra de cabeça no mundo dos ga...   
441  Depois das altas ligeiras, como na 5ª, vem a r...   
444  Clima pós 7 de setembro: em meio às incertezas...   
445  “Vamos lançar um NFT nos próximos meses que va...   
446     Digitalks terá presença de DAO 100% brasileira   

                                         PrimeiroParag  Target  \
4    O dólar abria o pregão desta segunda-feira (29...       0   
8    A Dommo Energia (DMMO3), antiga OGX de Eike Ba...       0   
12   A Ágora Investimentos, o BTG Pactual e o PagBa...       0   
14   A Agência Nacional de Energia Elétrica (Aneel)...       0   
15   Durante 40 minutos, Jair Bolsonaro (PL) foi en...       0   
..                                                 ...     ...   
439  O Magalu Games lançou seus primeiros três jogo...       0   
441  A virada da soja e do milho ontem, na segunda ...       0   
444  É fato consumado que vivemos as eleições mais ...       0   
445  Neste sábado (3) aconteceu o ViralCure Festiva...       0   
446  Cintia Ferreira, Fundadora da EVE e co-fundado...       0   

                                           Texto Final  
4    Ibovespa futuro opera em queda com aversão ao ...  
8    Prisma Capital se torna titular de 56% da Domm...  
12   Day Trade: Banco do Brasil (BBAS3), Marfrig (M...  
14   Diretores da Aneel aprovam novas tarifas de en...  
15   Bolsonaro no JN: centrão, pandemia, economia; ...  
..                                                 ...  
439  Magalu (MGLU3) entra de cabeça no mundo dos ga...  
441  Depois das altas ligeiras, como na 5ª, vem a r...  
444  Clima pós 7 de setembro: em meio às incertezas...  
445  “Vamos lançar um NFT nos próximos meses que va...  
446  Digitalks terá presença de DAO 100% brasileira...  

[134 rows x 4 columns]

In [60]:
train1 = train.loc[train['Target'] == 1,:]
train1

Titulo  \
1    IRB (IRBR3) é, de longe, a ação mais ‘shortead...   
2    5 ações que pagam bons dividendos para investi...   
3    Empiricus Investimentos escala 6 ações com div...   
6    BlackRock vê alta de ações de energia em merca...   
9    IRB (IRBR3) perde R$ 1 bilhão desde anúncio de...   
..                                                 ...   
433  Sem Wall Street, Ibovespa dispara puxado pelas...   
435  Day Trade: Petrobras (PETR4), Banco do Brasil ...   
437  Pechinchas da Bolsa e o que mais movimentou Co...   
440  Day Trade: 9 ações para comprar hoje e buscar ...   
448  Credores da chinesa Evergrande (EGRNF) montam ...   

                                         PrimeiroParag  Target  \
1    O IRB (IRBR3) é a ação com mais investidores s...       1   
2    Para setembro, a Órama Investimentos optou por...       1   
3    A Empiricus Investimentos, ex-Vitreo, incluiu ...       1   
6    O estrategista-chefe de investimentos da Black...       1   
9    O IRB  (IRBR3) perdeu R$ 1 bilhão em valor de ...       1   
..                                                 ...     ...   
433  O Ibovespa sobe forte na sessão desta segunda-...       1   
435  A Ágora Investimentos, o BTG Pactual, e o PagB...       1   
437  A Bolsa brasileira está barata. Isso não é nov...       1   
440  A Ágora Investimentos, o BTG Pactual, e o PagB...       1   
448  Fundos globais que investiram em títulos da ch...       1   

                                           Texto Final  
1    IRB (IRBR3) é, de longe, a ação mais ‘shortead...  
2    5 ações que pagam bons dividendos para investi...  
3    Empiricus Investimentos escala 6 ações com div...  
6    BlackRock vê alta de ações de energia em merca...  
9    IRB (IRBR3) perde R$ 1 bilhão desde anúncio de...  
..                                                 ...  
433  Sem Wall Street, Ibovespa dispara puxado pelas...  
435  Day Trade: Petrobras (PETR4), Banco do Brasil ...  
437  Pechinchas da Bolsa e o que mais movimentou Co...  
440  Day Trade: 9 ações para comprar hoje e buscar ...  
448  Credores da chinesa Evergrande (EGRNF) montam ...  

[166 rows x 4 columns]

In [61]:
train2 = train.loc[train['Target'] == 2,:]
train2

Titulo  \
0    Ibovespa: ALPA4 fecha mais um pregão em alta e...   
5    Ibovespa pode se agarrar em menor cautela e na...   
7    Apesar de dia mais fraco para o Ibovespa, CVCB...   
10   Small Caps têm rentabilidade 11,6% maior que o...   
13   IRB (IRBR3): Após a oferta de ações, é hora de...   
..                                                 ...   
438  Petrobras (PETR4) sobe 2% mesmo com corte no p...   
442  Banco do Brasil (BBAS3): Em novo preço-alvo, I...   
443  Oi (OIBR3) sobe 11% em uma semana e ultrapassa...   
447  Vale (VALE3), Santander (SANB11) e mais 3 açõe...   
449  Ibovespa: PCAR3 dispara mais de 8% e lidera as...   

                                         PrimeiroParag  Target  \
0    O Ibovespa (IBOV) fechou em queda de 1,09% nes...       2   
5    O Ibovespa (IBOV) seguiu ontem o mau humor das...       2   
7    O Ibovespa (IBOV) terminou o pregão fraco dest...       2   
10   A rentabilidade anual das Small Caps foi 11,6%...       2   
13   O IRB (IRBR3) recuperou parte das perdas com a...       2   
..                                                 ...     ...   
438  As ações da Petrobras (PETR3; PETR4) fecharam ...       2   
442  O Itaú BBA elevou o preço-alvo para a ação do ...       2   
443  A ação da Oi (OIBR3) deu um repique nos último...       2   
447  A Ativa Investimentos retirou as ações de Bras...       2   
449  Confira as maiores altas e baixas do Ibovespa ...       2   

                                           Texto Final  
0    Ibovespa: ALPA4 fecha mais um pregão em alta e...  
5    Ibovespa pode se agarrar em menor cautela e na...  
7    Apesar de dia mais fraco para o Ibovespa, CVCB...  
10   Small Caps têm rentabilidade 11,6% maior que o...  
13   IRB (IRBR3): Após a oferta de ações, é hora de...  
..                                                 ...  
438  Petrobras (PETR4) sobe 2% mesmo com corte no p...  
442  Banco do Brasil (BBAS3): Em novo preço-alvo, I...  
443  Oi (OIBR3) sobe 11% em uma semana e ultrapassa...  
447  Vale (VALE3), Santander (SANB11) e mais 3 açõe...  
449  Ibovespa: PCAR3 dispara mais de 8% e lidera as...  

[150 rows x 4 columns]

Faça aqui uma descrição do seu assunto e o que considerou como relevante ou não relevante na classificação das notícias (Target).

O Projeto 1 usará como fonte de dados o Portal do Bitcoin e o assunto das notícias será "ações". As nossas variáveis Target serão: muito relevante(2), relevante(1) e neutra(0).

Para muito relevante será analisado se: menciona diretamente o aumento ou queda na porcentagem de uma ação. Para relevante será considerado: não fala diretamente sobre as ações mas sim as pessoas que estão envolvidas com elas. Para neutras será considerado: empresas comprandos ações, em que não será mencionado as ações em si e os investidores.

___
### Montando um Classificador Naive-Bayes

Considerando apenas as mensagens da planilha Treinamento, ensine  seu classificador.

In [62]:
# Função que limpa e deixa todas as letras em minúsculas
import re 

def cleanup(text):
    punctuation = '(:)'
    pattern = re.compile(punctuation)
    text_subbed = re.sub(pattern, '', text)
    text_subbed = text_subbed.replace('\n',' ')
    text_subbed = text_subbed.replace('"',' ')
    text_subbed = text_subbed.replace("'",' ')
    text_subbed = text_subbed.lower()
    return text_subbed

In [63]:
# train.Titulo.to_string(index = False)

In [64]:
trainstr = train.Titulo.to_string(index = False) + train.PrimeiroParag.to_string(index = False)
teststr = test.Titulo.to_string(index = False) + test.PrimeiroParag.to_string(index = False)

train_clean = cleanup(trainstr)
test_clean = cleanup(teststr)

train_list = train_clean.split()
test_list = test_clean.split()

#train_clean

In [65]:
test_textof = test['Texto Final'].to_string(index = False)
test_textof = cleanup(test_textof)
teste = test_textof.split()

# test_textof

In [66]:
# Limpando os Textos finais de treinos
train0_tflimpo = cleanup(train0['Texto Final'].to_string(index=False))
train1_tflimpo = cleanup(train1['Texto Final'].to_string(index=False))
train2_tflimpo = cleanup(train2['Texto Final'].to_string(index=False))

In [67]:
# Lista com as palavras dos treinos
lista_t0 = list(train0_tflimpo.split())
lista_t1 = list(train1_tflimpo.split())
lista_t2 = list(train2_tflimpo.split())
lista_total = lista_t0 + lista_t1 + lista_t2
# Lista com todas as palavras das três listas mas sem repetição
lista_total_sem_repeticao = set(lista_t0 + lista_t1 + lista_t2) 

In [68]:
# Frequência absoluta das palavras em cada lista
freq_absoluta_t0 = pd.Series(lista_t0).value_counts()
freq_absoluta_t1 = pd.Series(lista_t1).value_counts()
freq_absoluta_t2 = pd.Series(lista_t2).value_counts()

In [69]:
# Frequência relativa das palavras em cada lista
freq_relativa_t0 = pd.Series(lista_t0).value_counts(True)
freq_relativa_t1 = pd.Series(lista_t1).value_counts(True)
freq_relativa_t2 = pd.Series(lista_t2).value_counts(True)

In [70]:
# Probabilidade das relevâncias
Prob_0 = len(lista_t0)/len(lista_total)
Prob_1 = len(lista_t1)/len(lista_total)
Prob_2 = len(lista_t2)/len(lista_total)

In [71]:
def NaiveBayes(titulo_noticia):
    probT0 = 1
    probT1 = 1
    probT2 = 1
    l0 = []
    titulo_noticia = titulo_noticia.split()
    for palavra in titulo_noticia:
        if palavra in lista_t0:
            probT0 *= freq_relativa_t0[palavra] / freq_relativa_t0.sum()
        elif palavra not in lista_t0:
            probT0 *= 1 / (len(lista_t0) + len(lista_total))
        if palavra in lista_t1:
            probT1 *= freq_relativa_t1[palavra] / freq_relativa_t1.sum()
        elif palavra not in lista_t1:
            probT1 *= 1 / (len(lista_t1) + len(lista_total))
        if palavra in lista_t2:
            probT2 *= freq_relativa_t2[palavra] / freq_relativa_t2.sum()
        elif palavra not in lista_t2:
            probT2 *= 1 / (len(lista_t2) + len(lista_total))
        probT0 *= Prob_0
        probT1 *= Prob_1
        probT2 *= Prob_2
    if probT0 > probT1 and probT0 > probT2:
        return 0
    elif probT1 > probT0 and probT1 > probT2:
        return 1
    elif probT2 > probT1 and probT2 > probT0:
        return 2

___
### Verificando a performance do Classificador

Agora você deve testar o seu classificador com a base de Testes.

In [98]:
test['Texto Final'] = cleanup(str(test['Texto Final']))

In [105]:
verificador = test.copy()

verificador['clean'] = verificador.Titulo.apply(cleanup)
verificador['comp'] = verificador.clean.apply(NaiveBayes)

verificador.head(5)

Titulo  \
0  Ibovespa (IBOV) vira para queda, com VALE3 ent...   
1  Ação da Braskem (BRKM3) avança com notícia sob...   
2  Compre commodities, se preocupe com recessão d...   
3  Veja as 5 novas ações escaladas pela XP para e...   
4  Respostas sobre Magazine Luiza (MGLU3) e mais ...   

                                       PrimeiroParag  Target  \
0  O Ibovespa (IBOV) operava sem forças nesta qua...       2   
1  As ações da Braskem (BRKM3) avançavam ao redor...       2   
2  O Goldman Sachs recomenda que os investidores ...       1   
3  A XP Investimentos repaginou sua carteira reco...       1   
4  Investidores não tiraram Magazine Luiza (MGLU3...       2   

                                         Texto Final  \
0  0      0      ibovespa (ibov) vira para queda,...   
1  0      0      ibovespa (ibov) vira para queda,...   
2  0      0      ibovespa (ibov) vira para queda,...   
3  0      0      ibovespa (ibov) vira para queda,...   
4  0      0      ibovespa (ibov) vira para queda,...   

                                               clean  comp  
0  ibovespa (ibov) vira para queda, com vale3 ent...     1  
1  ação da braskem (brkm3) avança com notícia sob...     1  
2  compre commodities, se preocupe com recessão d...     1  
3  veja as 5 novas ações escaladas pela xp para e...     1  
4  respostas sobre magazine luiza (mglu3) e mais ...     2

In [107]:
verd0 = verificador.loc[(verificador.comp == 0) & (verificador.Target == 0)]
verd1 = verificador.loc[(verificador.comp == 1) & (verificador.Target == 1)]
verd2 = verificador.loc[(verificador.comp == 2) & (verificador.Target == 2)]

fals0 = verificador.loc[(verificador.comp == 0) & (verificador.Target != 0)]
fals1 = verificador.loc[(verificador.comp == 1) & (verificador.Target != 1)]
fals2 = verificador.loc[(verificador.comp == 2) & (verificador.Target != 2)]

print(f'O percentual de verdadeiros negativos foi de: {100 * (verd0.shape[0] / verificador.shape[0]):.2f}%')
print(f'O percentual de falsos negativos foi de: {100 * (fals0.shape[0] / verificador.shape[0]):.2f}%')
print('\n')
print(f'O percentual de verdadeiros neutros foi de: {100 * (verd1.shape[0] / verificador.shape[0]):.2f}%')
print(f'O percentual de falsos neutros foi de: {100 * (fals1.shape[0] / verificador.shape[0]):.2f}%')
print('\n')
print(f'O percentual de verdadeiros positivos foi de: {100 * (verd2.shape[0] / verificador.shape[0]):.2f}%')
print(f'O percentual de falsos positivos foi de: {100 * (fals2.shape[0] / verificador.shape[0]):.2f}%')
print('\n')
print(f'A acurácia do modelo foi de: {100 * (verd0 + verd1 + verd2).shape[0] / test.shape[0]:.2f}%')

O percentual de verdadeiros negativos foi de: 7.33%
O percentual de falsos negativos foi de: 1.33%


O percentual de verdadeiros neutros foi de: 21.33%
O percentual de falsos neutros foi de: 28.67%


O percentual de verdadeiros positivos foi de: 20.33%
O percentual de falsos positivos foi de: 21.00%


A acurácia do modelo foi de: 49.00%


___
### Concluindo

#### Performance do classificador 
De acordo com o Teorema de Bayes o cálculo da probabilidade é dado por: $$P(A|B) = \frac{P(B|A) P(A)}{P(B)}$$  
Portanto, foi criada uma função ("NaiveBayes") que aplicasse tal raciocínio, a mesma proporcionou a aquisição dos novos valores de relevância de acordo com o algortimo do Classificador, cujos dados proporcionaram o cálculo dos percentuais de valores verdadeiros e falsos, além da acurácia.

O classificador automático obtido, que possui como propósito a classificação de notícias de acordo com a sua relevância sobre "ações", adquiriu uma acurácia de 49.00% quando utilizado com os dados da planilha Teste, um percentual de verdadeiros negativos de 7.33%, verdadeiros neutros de 21.33%, e de verdadeiros positivos de 20.33%.
Tal resultado se deve ao fato de que o classificador em questão, foi afetado pelos falsos neutros, visto que atribuem uma classificação improcedente ao classificador. Ademais, vale ressaltar que ao utilizar um database de menor escala, é esperado que suas classificações sejam mais arbitrárias, no entanto, se o mesmo fosse criado a partir de uma coleta de dados mais abundante, 


In [ ]:
# Dupla negação e sarcasmo

In [ ]:
# Plano de Expansão

In [ ]:
# Opcionais:

# Discorrer por que não posso alimentar minha base de Treinamento automaticamente usando o próprio classificador, aplicado a novas notícias.
# Propor diferentes cenários de uso para o classificador Naive-Bayes. Pense em outros cenários sem intersecção com este projeto.
# Sugerir e explicar melhorias reais no classificador com indicações concretas de como implementar (não é preciso codificar, mas indicar como fazer. Indique material de pesquisa sobre o assunto).

___
### Qualidade do Classificador a partir de novas separações das notícias entre Treinamento e Teste

Caso for fazer esse item do Projeto

___
## Aperfeiçoamento:

Trabalhos que conseguirem pelo menos conceito B vão evoluir em conceito dependendo da quantidade de itens avançados:

* IMPLEMENTOU outras limpezas e transformações que não afetem a qualidade da informação contida nas notícias. Ex: stemming, lemmatization, stopwords
* CONSIDEROU mais de duas caterogias na variável Target e INCREMENTOU a quantidade de notícias, mantendo pelo menos 250 notícias por caterogia (OBRIGATÓRIO PARA TRIOS, sem contar como item avançado)
* Para Target com duas categorias: CRIOU pelo menos quatro categorias intermediárias de relevância baseadas na probabilidade: ex.: muito relevante, relevante, neutro, irrelevante, muito irrelevante
* EXPLICOU porquê não pode usar o próprio classificador para gerar mais amostras de treinamento
* PROPÔS diferentes cenários para Naïve Bayes fora do contexto do projeto (pelo menos dois cenários, exceto aqueles já apresentados em sala pelos professores: por exemplo, filtro de spam)
* SUGERIU e EXPLICOU melhorias reais com indicações concretas de como implementar (indicar como fazer e indicar material de pesquisa)
* FEZ o item Qualidade do Classificador a partir de novas separações das Notícias entre Treinamento e Teste descrito no enunciado do projeto (OBRIGATÓRIO para conceitos A ou A+)

___
## Referências

[Naive Bayes and Text Classification](https://arxiv.org/pdf/1410.5329.pdf)  **Mais completo**

[A practical explanation of a Naive Bayes Classifier](https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/) **Mais simples**